File: BFS_DFS/Problems/BJ_1182.py
# 문제 정의: 부분 수열의 합
# 문제 유형: 백트래킹
# 수준: 실버 2
# 접근: DFS(Backtracking)

N, S = map(int, input().split())
lst = list(map(int, input().split()))
cnt = 0 
def dfs(i, s):
  global cnt
  if i >= N:
    return
  
  if s+lst[i] == S:
    cnt += 1

  # 두 번 찾아서 모든 가능성을 탐색
  dfs(i+1, s+lst[i])
  dfs(i+1, s)  

dfs(0, 0)
print(cnt)

File: BFS_DFS/Problems/Backtracking_Basics/BJ_1182.java
// Backtracking | 부분 수열 중 특정 합을 만족하는 합의 개수
// 분기를 잘 타는 게 중요!

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BJ_1182 {
  static int cnt, N, S;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    N = Integer.parseInt(st.nextToken());
    S = Integer.parseInt(st.nextToken());

    int arr[] = new int[N];
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < arr.length; i++) {
      arr[i] = Integer.parseInt(st.nextToken());
    }

    dfs(arr, 0, 0);

    System.out.println(cnt - (S == 0 ? 1 : 0));
  }

  static void dfs(int[] arr, int idx, int sum) {
    if (idx == N) {
      if (sum == S) {
        cnt++;
      }
      return;
    }

    dfs(arr, idx + 1, sum + arr[idx]);
    dfs(arr, idx + 1, sum);
  }
}
